Library/Homebrew/cmd/update.sh

```diff
@@ -378,9 +385,14 @@ EOS
   if [[ -n "$HOMEBREW_UPDATED" || -n "$HOMEBREW_UPDATE_FAILED" ]]
   then
     brew update-report "$@"
-    return $?
+    exitcode=$?
+    unlock update
+    return $exitcode
   elif [[ -z "$HOMEBREW_UPDATE_PREINSTALL" ]]
   then
     echo "Already up-to-date."
+    unlock update
+  else
+    unlock update
```

Comment:
I think it'd be more robust to `trap "unlock update" EXIT` rather than manually adding an `unlock` at each exit point.